toolchain/parse/typed_nodes.h

```diff
@@ -154,28 +161,36 @@ using SelfTypeNameExpr =
 using BaseName =
     LeafNode<NodeKind::BaseName, Lex::BaseTokenIndex, NodeCategory::MemberName>;
 
-// An unqualified name and optionally a following sequence of parameters.
-// For example, `A`, `A(n: i32)`, or `A[T:! type](n: T)`.
-struct NameAndParams {
-  IdentifierNameId name;
+// A name qualifier with parameters, such as `A(T:! type).` or `A[T:! type](N:!
+// T).`.
+struct NameQualifierWithParams {
+  static constexpr auto Kind = NodeKind::NameQualifierWithParams.Define(
+      {.category = NodeCategory::NameQualifier,
+       .bracketed_by = IdentifierNameBeforeParams::Kind});
+
+  IdentifierNameBeforeParamsId name;
   std::optional<ImplicitParamListId> implicit_params;
   std::optional<TuplePatternId> params;
+  Lex::PeriodTokenIndex token;
 };
 
-// A name qualifier: `A.`, `A(T:! type).`, or `A[T:! type](N:! T).`.
-struct NameQualifier {
-  static constexpr auto Kind =
-      NodeKind::NameQualifier.Define({.bracketed_by = IdentifierName::Kind});
+// A name qualifier without parameters, such as `A.`.
+struct NameQualifierWithoutParams {
+  static constexpr auto Kind = NodeKind::NameQualifierWithoutParams.Define(
+      {.category = NodeCategory::NameQualifier,
+       .bracketed_by = IdentifierNameNotBeforeParams::Kind});
 
-  NameAndParams name_and_params;
+  IdentifierNameNotBeforeParamsId name;
   Lex::PeriodTokenIndex token;
 };
 
 // A complete name in a declaration: `A.C(T:! type).F(n: i32)`.
 // Note that this includes the parameters of the entity itself.
 struct DeclName {
-  llvm::SmallVector<NameQualifierId> qualifiers;
-  NameAndParams name_and_params;
+  llvm::SmallVector<AnyNameQualifierId> qualifiers;
+  AnyNonExprIdentifierNameId name;
+  std::optional<ImplicitParamListId> implicit_params;
+  std::optional<TuplePatternId> params;
```

Comment:
Had you considered, or what do you think of:

```suggestion
  struct NameWithParams {
    IdentifierNameBeforeParams name;
    std::optional<ImplicitParamListId> implicit_params;
    std::optional<TuplePatternId> params;
  };
  llvm::SmallVector<AnyNameQualifierId> qualifiers;
  std::optional<IdentifierNameNotBeforeParams> name_without_params;
  std::optional<NameWithParams> name_with_params;
```

(with the above structure, can `params` be non-optional or do we have cases with only implicit params?)

Note, if you like this, a similar restructuring in `Alternative` might make sense -- `parameters` could also be non-optional there. If you switched both, it might remove the need for the `NonExprIdentifierName` category -- but I'm really just bringing this up as a possible way to have tree verification enforce that params are present only when `IdentifierNameBeforeParams` is in the tree (even though it doesn't enforce exclusivity of name_without_params and name_with_params).

Reply:
I'd prefer not to do that, because as a reader I find it confusing to use a pair of `std::optional`s to represent mutually exclusive alternatives. Something like `NodeIdOneOf<IdentifierNameNotBeforeParams, NameWithParams>` would be ideal, except that `NameWithParams` isn't a node kind, and we can't make it one because there's no token for it to attach to.

> (with the above structure, can `params` be non-optional or do we have cases with only implicit params?)

I believe the toolchain currently doesn't support omitting the explicit parameter list, but @josh11b mentioned there's been discussion of making it optional in some cases, so it seems better to keep our options open.

Reply:
I believe the lambda proposal added that.